{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aef45646",
   "metadata": {},
   "source": [
    "import gffutils\n",
    "fn = gffutils.example_filename('C:/Users/clime/OneDrive/Documents/gencode.v41.annotation.gff3')\n",
    "print(open(fn).read()) \n",
    "#do for small data like 1 gene with 3 transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "b8871c61",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        start       end                    Transcript_ID    Type  \\\n",
      "37   43063333  43063373  transcript_id=ENST00000352993.7  ID=CDS   \n",
      "27   43099775  43099880  transcript_id=ENST00000352993.7  ID=CDS   \n",
      "26   43104122  43104261  transcript_id=ENST00000352993.7  ID=CDS   \n",
      "25   43104868  43104956  transcript_id=ENST00000352993.7  ID=CDS   \n",
      "24   43106456  43106533  transcript_id=ENST00000352993.7  ID=CDS   \n",
      "..        ...       ...                              ...     ...   \n",
      "232  43076488  43076611  transcript_id=ENST00000700182.1  ID=CDS   \n",
      "228  43095846  43095922  transcript_id=ENST00000700182.1  ID=CDS   \n",
      "283  43110535  43110580  transcript_id=ENST00000700183.1  ID=CDS   \n",
      "282  43115726  43115779  transcript_id=ENST00000700183.1  ID=CDS   \n",
      "281  43124017  43124096  transcript_id=ENST00000700183.1  ID=CDS   \n",
      "\n",
      "                CDS_ID  \n",
      "37   ENST00000352993.7  \n",
      "27   ENST00000352993.7  \n",
      "26   ENST00000352993.7  \n",
      "25   ENST00000352993.7  \n",
      "24   ENST00000352993.7  \n",
      "..                 ...  \n",
      "232  ENST00000700182.1  \n",
      "228  ENST00000700182.1  \n",
      "283  ENST00000700183.1  \n",
      "282  ENST00000700183.1  \n",
      "281  ENST00000700183.1  \n",
      "\n",
      "[324 rows x 5 columns]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import  natsort\n",
    "\n",
    "df = pd.read_fwf(\"C:/Users/clime/Downloads/BRCA1_CDS_annotations.txt\", header = None)\n",
    "\n",
    "df2 = df.loc[:,(3,4,8)] #parses out columns\n",
    "df2.columns = df2.columns.astype(str) #changes column int to column str\n",
    "df2.columns.values[0] = 'start' #changes column name from 3 to start\n",
    "df2.columns.values[1] = 'end'\n",
    "\n",
    "#print(df2)\n",
    "#print(df3)\n",
    "\n",
    "df4 = pd.concat([df2[['start','end']], df2['8'].str.split(';', expand = True)], axis = 1) #splits up the 9th column\n",
    "#axis defaults to rows with axis=0, axis=1=columns\n",
    "df4.columns = df4.columns.astype(str) #changes mixed column types to all str\n",
    "                    \n",
    "#print(df4)\n",
    "\n",
    "np.savetxt(r'C:/Users/clime/Downloads/BRCA1_CDS_ann', df4_1.values, fmt = '%s')\n",
    "\n",
    "df4_1 = df4.sort_values('0', key = natsort.natsort_keygen())\n",
    "#print(df4_1)\n",
    "\n",
    "#np.savetxt(r'C:/Users/clime/Downloads/BRCA1_CDS_ann_1', df4_1.values, fmt = '%s')\n",
    "\n",
    "df5 = df4.loc[:,('start','end','3','0')] #parses for CDS ID\n",
    "\n",
    "df6 = pd.concat([df5[['start','end',\"3\"]], df5['0'].str.split(':', expand = True)], axis = 1)\n",
    "df6.columns = df6.columns.astype(str)\n",
    "#df6 = df6.loc[:,('start','end','1')] #only shows CDS ID\n",
    "#concat and df4.loc column was 0, using 9 temp str.split was \":\" now '='\n",
    "#print(df6)\n",
    "\n",
    "df7 = df6.sort_values('1', key = natsort.natsort_keygen())\n",
    "df7.columns.values[2] = 'Transcript_ID'\n",
    "df7.columns.values[3] = 'Type'\n",
    "df7.columns.values[4] = 'CDS_ID'\n",
    "df7.columns = df7.columns.astype(str)\n",
    "print(df7)\n",
    "\n",
    "start = df7[\"start\"]\n",
    "end = df7[\"end\"]\n",
    "transcript = [\"Transcript_ID\"]\n",
    "zero = df7[\"Type\"]\n",
    "CDS_ID = df7[\"CDS_ID\"]\n",
    "\n",
    "df7.to_csv('C:/Users/clime/Downloads/BRCA1_CDS_extract_ann_7.csv', header=[start,end,transcript,zero,CDS_ID], ) \n",
    "    \n",
    "#np.savetxt(r'C:/Users/clime/Downloads/BRCA1_CDS_extract_ann_7.', df7.values, fmt = '%s')\n",
    "#np.savetxt(r'C:/Users/clime/Downloads/BRCA1_CDS_locations', df2.values, fmt='%s') \n",
    "#df3.to_csv(r'C:/Users/clime/Downloads/BRCA1_CDS_column9.txt', header=None, \n",
    "          #index=None, sep=';', mode='a') allows to write into a csv readable txt file or csv if remove .txt\n",
    "    \n",
    "#pd.read_fwf() fixed-width lines can help with breaking the file with a fixed width space but ; and , in columns\n",
    "#np.savetxt(r'~/filename.txt', df.values, fmt='%s') to write txt file fmt='%s' for string format instead of delimiter\n",
    "#df.to_csv(r'c:\\data\\pandas.txt', header=None, index=None, sep=' ', mode='a') to write csv file\n",
    "\n",
    "s = list(df7['CDS_ID'])\n",
    "#print(s)\n",
    "with open('C:/Users/clime/Downloads/BRCA1_CDS_ID.csv', 'w') as f:\n",
    "    for entry in s:\n",
    "        f.write(entry + \"\\n\")\n",
    "        \n",
    "#print(list(set(s)))\n",
    "\n",
    "with open('C:/Users/clime/Downloads/BRCA1_CDS_ID_Nodupes.csv', 'w') as f:\n",
    "    for entry in list(set(s)):\n",
    "        f.write(entry + '\\n')\n",
    "        \n",
    "df8 = df4.loc[:,('start','end','7')]\n",
    "df9 = pd.concat([df8[['start','end']], df8['7'].str.split('=', expand = True)], axis = 1)\n",
    "df9.columns = df9.columns.astype(str)\n",
    "#print(df9)\n",
    "s_1 = list(df9['1'])\n",
    "#print(s)\n",
    "with open('C:/Users/clime/Downloads/BRCA1_CDS_ID_1.csv', 'w') as f:\n",
    "    for entry in s_1:\n",
    "        f.write(entry + \"\\n\")\n",
    "        \n",
    "#print(list(set(s)))\n",
    "\n",
    "with open('C:/Users/clime/Downloads/BRCA1_CDS_ID_Nodupes_1.csv', 'w') as f:\n",
    "    for entry in list(set(s_1)):\n",
    "        f.write(entry + '\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
