{"cells":[{"source":"# Load packages\nlibrary(Biostrings)\n\n# Check the alphabet of the zikaVirus\nalphabet(zikaVirus)\n\n# Check the alphabetFrequency of the zikaVirus\nalphabetFrequency(zikaVirus)\n\n# Check alphabet of the zikaVirus using baseOnly = TRUE\nalphabet(zikaVirus, baseOnly = TRUE)\n\n# Unlist the set, select the first 21 letters, and assign to dna_seq\ndna_seq <- subseq(unlist(zikaVirus), end = 21)\ndna_seq\n\n# Transcribe dna_seq into an RNAString object and print it\nrna_seq <- RNAString(dna_seq) \nrna_seq\n\n# Translate rna_seq into an AAString object and print it\naa_seq <- translate(rna_seq)\naa_seq\n\n# Unlist the set, select the first 21 letters, and assign to dna_seq\ndna_seq <- subseq(unlist(zikaVirus), end = 21)\ndna_seq\n\n# Transcribe and translate dna_seq into an AAString object and print it\naa_seq <- translate(dna_seq)\naa_seq\n\n# to create a new set from a single sequencezikaSet <- DNAStringSet(zikaVirus_seq, start =c(1,101,201), end =c(100,200,300))\n\n# Print rnaframesZikaSet\nrnaframesZikaSet\n\n# Translate rnaframesZikaSet\nAAzika6F <- translate(rnaframesZikaSet)\nAAzika6F\n\n# Count NS5 protein matches in AAzika6F, allowing 15 mismatches\nvcountPattern(pattern = NS5, subject = AAzika6F, max.mismatch = 15)\n\n# Use vmatchPattern() with the set\nvmatchPattern(pattern = ns5, subject = selectedSet, max.mismatch = 15)\n\n# Load IRanges package\nlibrary(IRanges)\n\n# IRnum1: start - vector 1 through 5, end - 100  \nIRnum1 <- IRanges(start = c(1:5), end = 100)\n\n# IRnum2: end - 100, width - 89 and 10\nIRnum2 <- IRanges(end = 100, width = c(89, 10))\n\n# IRlog1: start = Rle(c(F, T, T, T, F, T, T, T)))\nIRlog1 <- IRanges(start = Rle(c(F, T, T, T, F, T, T, T)))\n\n# Print objects in a list\nprint(list(IRnum1 = IRnum1, IRnum2 = IRnum2, IRlog1 = IRlog1))\n\n# Create the first sequence seq_1\nseq_1 <- IRanges(start = 10, end = 37)\n\n# Create the second sequence seq_2\nseq_2 <- IRanges(start = c(5, 35, 50),\n                 end = c(12, 39, 61),\n                 names = LETTERS[1:3])\n\n# Check the width of seq_1 and seq_2\nlengths(seq_1)\nlengths(seq_2)\n#OR with width()\n\n# Load GenomicRanges package\nlibrary(GenomicRanges)\n\n# Print seq_intervals\nseq_intervals\n\n# Create myGR\nmyGR <- as(seq_intervals, \"GRanges\")\n\n# Print myGR\nmyGR\n\n#accessors of GRanges\nseqnames(gr)\nranges(gr)\nmcols(gr)\ngenome(gr)\nseqinfo(gr)\n\n#to see all of them\nmethods(class = \"GRanges\")\n\n# Load human reference genome hg38\nlibrary(TxDb.Hsapiens.UCSC.hg38.knownGene)\n\n# Assign hg38 to hg, then print it\nhg <- TxDb.Hsapiens.UCSC.hg38.knownGene\nhg\n\n# Extract all positive stranded genes in chromosome X, assign to hg_chrXgp, then sort it\nhg_chrXgp <- genes(hg, filter = list(tx_chrom = 'chrX', tx_strand = \"+\"))\nsort(hg_chrXgp)\n\n#overlapsAny() between the target ABCD1 and the subject hg_chrX\noverlapsAny(ABCD1, hg_chrX)\n#returns logical\n\n# Store the overlapping range in rangefound\nrangefound <- subsetByOverlaps(hg_chrX, ABCD1)\n\n# Print names of rangefound\nnames(rangefound)\n\n# Print the gene of interest \nABCD1\n\n# Print rangefound\nrangefound\n\n# Load the human transcripts DB to hg\nlibrary(TxDb.Hsapiens.UCSC.hg38.knownGene)\nhg <- TxDb.Hsapiens.UCSC.hg38.knownGene\n\n# Prefilter chromosome X \"chrX\" using seqlevels()\nseqlevels(hg) <- c('chrX')\n\n# Get all transcripts by gene and print it\nhg_chrXt <- transcriptsBy(hg, by = 'gene')\nhg_chrXt\n\n# Select gene `215` from the hg_chrXt\nhg_chrXt$`215`\n\n# Unlist hg_ChrX and save result as myGR\nmyGR <- unlist(hg_ChrX)\n\n# Compare classes of hg_ChrX and myGR\nclass(hg_ChrX)\nclass(myGR)\n\n# Compare length of hg_ChrX and myGR\nwidth(hg_ChrX)\nwidth(myGR)\n\n# Load ShortRead\nlibrary(ShortRead)\n\n# Print fqsample\nfqsample\n\n# Check class of fqsample\nclass(fqsample)\n\n# Check class sread fqsample\nclass(sread(fqsample))\n\n# Check ids of fqsample\nid(fqsample)\n\n# Load ShortRead\nlibrary(ShortRead)\n\n# Set a seed for sampling\nset.seed(1234)\n\n# Use FastqSampler with f and select 100 reads\nfs <- FastqSampler(con = f, n = 100)\n\n# Generate new sample yield\nmy_sample <- yield(fs)\n\n# Print my_sample\nmy_sample\n\n# load ShortRead\nlibrary(ShortRead)\n\n# Check quality\nquality(fqsample)\n\n# Check encoding\nencoding(quality(fqsample))\n\n# load ShortRead\nlibrary(ShortRead)\n\n# Check quality\nquality(fqsample)\n\n# Check encoding of quality\nencoding(quality(fqsample))\n\n# Check baseQuality\nqaSummary[['baseQuality']]\n\n# Glimpse nucByCycle\nglimpse(nucByCycle)\n\n# Create a line plot of cycle vs. count\nnucByCycle %>% \n  # Gather the nucleotide letters in alphabet and get a new count column\n  pivot_longer(-cycle, names_to = 'alphabet', values_to = 'count') %>% \n  ggplot(aes(x = cycle, y =  count, color = alphabet)) +\n  geom_line(size = 0.5 ) +\n  labs(y = \"Frequency\") +\n  theme_bw() +\n  theme(panel.grid.major.x = element_blank())\n\n# Load package ShortRead\nlibrary(ShortRead)\n\n# Check class of fqsample\nclass(fqsample)\n\n# Filter reads into selectedReads using myStartFilter\nselectedReads <- fqsample[myStartFilter(fqsample)]\n\n# Check class of selectedReads\nclass(selectedReads)\n\n# Check detail of selectedReads\ndetail(selectedReads)\n\n#To keep those reads that are not duplicated \nmydReads[srduplicated(mydReads) == FALSE]\n\n# Check reads of fqsample\nsread(fqsample)\n\n# Create myFil using polynFilter\nmyFil <- polynFilter(threshold = 3, nuc = c(\"A\"))\n\n# Apply your filter to fqsample\nfilterCondition <- myFil(fqsample)\n\n# Use myFil with fqsample\nfilteredSequences <- fqsample[filterCondition]\n\n# Check reads of filteredSequences\nsread(filteredSequences)\n\n# Load package Rqc\nlibrary(Rqc)\n\n# Average per cycle quality plot\nrqcCycleAverageQualityPlot(qa)\n\n# Load package Rqc\nlibrary(Rqc)\n\n# Average per cycle quality plot with white background\nrqcCycleAverageQualityPlot(qa) + theme_minimal()","metadata":{},"id":"forbidden-likelihood","cell_type":"code","execution_count":null,"outputs":[]},{"source":"# Explore the first six observations of smoc2_rawcounts\nhead(smoc2_rawcounts)\n\n# Explore the structure of smoc2_rawcounts\nstr(smoc2_rawcounts)\n\n\n# Use the match() function to reorder the columns of the raw counts\nreorder_idx <- match(rownames(smoc2_metadata), colnames(smoc2_rawcounts))\n\n\n# Reorder the columns of the count data\nreordered_smoc2_rawcounts <- smoc2_rawcounts[ , reorder_idx]\n\n# Create a DESeq2 object\ndds_smoc2 <- DESeqDataSetFromMatrix(countData = reordered_smoc2_rawcounts,\n                              colData = smoc2_metadata,\n                              design = ~ condition)\n\n# Transform the normalized counts \nvsd_smoc2 <- vst(dds_smoc2, blind = TRUE)\n\n# Extract the matrix of transformed counts\nvsd_mat_smoc2 <- assay(vsd_smoc2)\n\n# Compute the correlation values between samples\nvsd_cor_smoc2 <- cor(vsd_mat_smoc2) \n\n# Plot the heatmap\npheatmap(vsd_cor_smoc2, annotation = select(smoc2_metadata, condition))\n\n#The biological replicates cluster together and the samples in different conditions cluster separately. There are no outliers or samples with low correlation values relative to all other samples.\n\n# Transform the normalized counts \nvsd_smoc2 <- vst(dds_smoc2, blind = TRUE)\n\n# Plot the PCA of PC1 and PC2\nplotPCA(vsd_smoc2, intgroup='condition')\n\n# Create DESeq2 object\ndds_smoc2 <- DESeqDataSetFromMatrix(countData = reordered_smoc2_rawcounts,\n                 colData = smoc2_metadata,\n                 design = ~ condition)\n\n# Run the DESeq2 analysis\ndds_smoc2 <- DESeq(dds_smoc2)\n\n# Plot dispersions\nplotDispEsts(dds_smoc2)\n\n# Extract the results of the differential expression analysis\nsmoc2_res <- results(dds_smoc2, \n                condition = fibrosis, \n                alpha = 0.05)\n\n# Shrink the log2 fold change estimates to be more accurate\nsmoc2_res <- lfcShrink(dds_smoc2, \n                    contrast =  c('condition', 'fibrosis', 'normal'),\n                    res = smoc2_res)\n\n# Extract results\nsmoc2_res <- results(dds_smoc2, \n                contrast = c('condition', 'fibrosis', 'normal'), \n                alpha = 0.05, \n                lfcThreshold = 0.32)\n\n# Shrink the log2 fold changes\nsmoc2_res <- lfcShrink(dds_smoc2, \n                    contrast = c('condition', 'fibrosis', 'normal'), \n                    res = smoc2_res)\n\n# Subset normalized counts to significant genes\nsig_norm_counts_smoc2 <- normalized_counts_smoc2[rownames(smoc2_res_sig), ]\n\n# Choose heatmap color palette\nheat_colors <- brewer.pal(n = 6, name = \"YlOrRd\")\n\n# Plot heatmap\npheatmap(sig_norm_counts_smoc2, \n         color = heat_colors, \n         cluster_rows = TRUE, \n         show_rownames = FALSE,\n         annotation = select(smoc2_metadata, condition), \n         scale = 'row')\n\n# Check that all of the samples are in the same order in the metadata and count data\nall(rownames(all_metadata) %in% colnames(all_rawcounts))\n\n# DESeq object to test for the effect of fibrosis regardless of genotype\ndds_all <- DESeqDataSetFromMatrix(countData = all_rawcounts,\n                        colData = all_metadata,\n                        design = ~ genotype + condition)\n\n# DESeq object to test for the effect of genotype on the effect of fibrosis                        \ndds_complex <- DESeqDataSetFromMatrix(countData = all_rawcounts,\n                                colData = all_metadata,\n                                design = ~ genotype + condition + genotype:condition)\n\n# Log transform counts for QC\nvsd_all <- vst(dds_all, blind = TRUE)\n\n# Create heatmap of sample correlation values\nvsd_all %>% \n        assay() %>%\n        cor() %>%\n        pheatmap(annotation =  select(all_metadata, c(\"condition\", \"genotype\")))\n        \n# Create the PCA plot for PC1 and PC2 and color by condition \nplotPCA(vsd_all, intgroup = \"condition\")\n\n# Create the PCA plot for PC1 and PC2 and color by genotype    \nplotPCA(vsd_all, intgroup = \"genotype\")\n\n# Select significant genese with padj < 0.05\nsmoc2_sig <- subset(res_all, padj < 0.05) %>%\n  \t\t\t\tdata.frame() %>%\n  \t\t\t\trownames_to_column(var = 'geneID')\n\n# Extract the top 6 genes with padj values\nsmoc2_sig %>%\n\tarrange(padj) %>%\n\tselect(geneID, padj) %>%\n\thead()\n","metadata":{},"cell_type":"code","id":"c1018538-e54b-4ae2-a2f0-c7d647a841b5","execution_count":null,"outputs":[]},{"source":"# Create a boxplot of the first gene in the expression matrix\nboxplot(x[1, ] ~ p[, \"Disease\"], main = f[1, \"symbol\"])\n\n# Load package\nlibrary(Biobase)\n\n# Create ExpressionSet object\neset <- ExpressionSet(assayData = x,\n                      phenoData = AnnotatedDataFrame(p),\n                      featureData = AnnotatedDataFrame(f))\n\n# View the number of features (rows) and samples (columns)\ndim(eset)\n\n# Subset to only include the first 10 samples (columns)\neset_sub <- eset[, 10]\n\n# Check the dimensions of the subset\ndim(eset_sub)\n\n# Create a boxplot of the 1000th gene in eset_sub\nboxplot(exprs(eset_sub)[1000, ] ~ pData(eset_sub)[, \"Disease\"],\n        main = fData(eset_sub)[1000, \"symbol\"])\n\n# Create design matrix for leukemia study\ndesign <- model.matrix(SampleID ~ Disease, data = pData(eset))\n\n# Count the number of samples modeled by each coefficient\ncolSums(design)\n\n# Create design matrix for leukemia study\ndesign <- model.matrix(SampleID ~ Disease, data = pData(eset))\n\n# Count the number of samples modeled by each coefficient\ncolSums(design)\n\n#if need help learning how to specift the model.matrix() design look at the data being used with ex:\nsummary() #OR\nhead() #OR\ncolnames() #ETC\n\n# Load package\nlibrary(limma)\n\n# Fit the model\nfit <- lmFit(eset, design)\n\n# Calculate the t-statistics\nfit <- eBayes(fit)\n\n# Summarize results\nresults <- decideTests(fit[, \"Diseasestable\"])\nsummary(results)\n\n# Create design matrix with no intercept\ndesign <- model.matrix(~0 + Disease, data = pData(eset))\n\n# Count the number of samples modeled by each coefficient\ncolSums(design)\n\n# Load package\nlibrary(limma)\n\n# Create a contrasts matrix\ncm <- makeContrasts(status = Diseaseprogres. - Diseasestable,\n                    levels = design)\n\n# View the contrasts matrix\ncm\n\n# Load package\nlibrary(limma)\n\n# Fit the model\nfit <- lmFit(eset, design)\n\n# Fit the contrasts\nfit2 <- contrasts.fit(fit, contrasts = cm)\n\n# Calculate the t-statistics for the contrasts\nfit2 <- eBayes(fit2)\n\n# Summarize results\nresults <- decideTests(fit2)\nsummary(results)\n\n# Create design matrix with no intercept\ndesign <- model.matrix(~0 + oxygen, data = pData(eset))\n\n# Count the number of samples modeled by each coefficient\ncolSums(design)\n\n# Load package\nlibrary(limma)\n\n# Create a contrasts matrix\ncm <- makeContrasts(ox05vox01 = oxygenox05 - oxygenox01,\n                    ox21vox01 = oxygenox21 - oxygenox01,\n                    ox21vox05 = oxygenox21 - oxygenox05,\n                    levels = design)\n\n# View the contrasts matrix\ncm\n\n# Load package\nlibrary(limma)\n\n# Fit the model\nfit <- lmFit(eset, design)\n\n# Fit the contrasts\nfit2 <- contrasts.fit(fit, contrasts = cm)\n\n# Calculate the t-statistics for the contrasts\nfit2 <- eBayes(fit2)\n\n# Summarize results\nresults <- decideTests(fit2)\nsummary(results)\n\n# Create single variable\ngroup <- with(pData(eset), paste(type, water, sep = \".\"))\ngroup <- factor(group)\n\n# Create design matrix with no intercept\ndesign <- model.matrix(~0 + group)\ncolnames(design) <- levels(group)\n\n# Count the number of samples modeled by each coefficient\ncolSums(design)\n\n# Load package\nlibrary(limma)\n\n# Create a contrasts matrix\ncm <- makeContrasts(type_normal = nm6.normal - dn34.normal,\n                    type_drought = nm6.drought - dn34.drought,\n                    water_nm6 = nm6.drought - nm6.normal,\n                    water_dn34 = dn34.drought - dn34.normal,\n                    interaction = (nm6.drought - nm6.normal) - (dn34.drought - dn34.normal),\n                    levels = design)\n\n# View the contrasts matrix\ncm\n\nlibrary(limma)\n\n# Create new ExpressionSet to store normalized data\neset_norm <- eset_raw\n\n# View the distribution of the raw data\nplotDensities(eset_norm, legend = FALSE)\n\n# Log tranform\nexprs(eset_norm) <- log(exprs(eset_norm))\nplotDensities(eset_norm, legend = FALSE)\n\n# Quantile normalize\nexprs(eset_norm) <- normalizeBetweenArrays(exprs(eset_norm))\nplotDensities(eset_norm, legend = FALSE)\n\nlibrary(limma)\n\n# Create new ExpressionSet to store filtered data\neset <- eset_norm\n\n# View the normalized gene expression levels\nplotDensities(eset, legend = FALSE); abline(v = 5)\n\n# Determine the genes with mean expression level greater than 5\nkeep <- rowMeans(exprs(eset)) > 5\nsum(keep)\n\n# Filter the genes\neset <- eset[keep,]\nplotDensities(eset, legend = FALSE)\n\n# Load package\nlibrary(limma)\n\n# Plot principal components labeled by treatment\nplotMDS(eset, labels = pData(eset)[, \"treatment\"], gene.selection = \"common\")\n\n# Plot principal components labeled by batch\nplotMDS(eset, labels = pData(eset)[, \"batch\"], gene.selection = \"common\")\n\n# Load package\nlibrary(limma)\n\n# Remove the batch effect\nexprs(eset) <- removeBatchEffect(eset, batch = pData(eset)[,\"batch\"])\n\n# Plot principal components labeled by treatment\nplotMDS(eset, labels = pData(eset)[, 'treatment'], gene.selection = 'common')\n\n# Plot principal components labeled by batch\nplotMDS(eset, labels = pData(eset)[, 'batch'], gene.selection = 'common')\n\n# Obtain the summary statistics for every gene\nstats <- topTable(fit2, number = nrow(fit2), sort.by = 'none')\n\n# Plot a histogram of the p-values\nhist(stats[, 'P.Value'])\n\n# Create a volcano plot. Highlight the top 5 genes\nvolcanoplot(fit2, highlight = 5, names = fit2$genes[, 'symbol'])\n\n# Extract the entrez gene IDs\nentrez <- fit2$genes[, 'entrez']\n\n# Test for enriched KEGG Pathways\nenrich_kegg <- kegga(fit2, geneid = entrez, species = 'Hs')\n\n# View the top 20 enriched KEGG pathways\ntopKEGG(enrich_kegg)\n\n# Extract the entrez gene IDs\nentrez <- fit2$genes[, \"entrez\"]\n\n# Test for enriched GO categories\nenrich_go <- goana(fit2[1:500, ], geneid = entrez[1:500], species = 'Hs')\n\n# View the top 20 enriched GO Biological Processes\ntopGO(enrich_go, ontology = 'BP')\n\n# Create a new ExpressionSet to store the processed data\neset <- eset_raw\n# Log transform\nexprs(eset) <- log(exprs(eset))\nplotDensities(eset,  group = pData(eset)[, 'genotype'], legend = \"topright\")\n# Quantile normalize\nexprs(eset) <- normalizeBetweenArrays(exprs(eset))\nplotDensities(eset,  group = pData(eset)[, 'genotype'], legend = \"topright\")\n# Determine the genes with mean expression level greater than 0\nkeep <- rowMeans(exprs(eset)) > 0\nsum(keep)\n# Filter the genes\neset <- eset[keep,]\nplotDensities(eset, group = pData(eset)[, 'genotype'], legend = \"topright\")\n\n# Find the row which contains Top2b expression data\ntop2b <- which(fData(eset)[, 'symbol'] == \"Top2b\")\n\n# Plot Top2b expression versus genotype\nboxplot(exprs(eset)[top2b, ] ~ pData(eset)[, 'genotype'],\n        main = fData(eset)[top2b, ])\n\n# Plot principal components labeled by genotype\nplotMDS(eset, labels = pData(eset)[, 'genotype'], gene.selection = \"common\")\n\n# Plot principal components labeled by treatment\nplotMDS(eset, labels = pData(eset)[, 'treatment'], gene.selection = \"common\")\n\n# Create single variable\ngroup <- with(pData(eset), paste(genotype, treatment, sep = \".\"))\ngroup <- factor(group)\n\n# Create design matrix with no intercept\ndesign <- model.matrix(~0 + group)\ncolnames(design) <- levels(group)\n\n# Count the number of samples modeled by each coefficient\ncolSums(design)\n\n# Create a contrasts matrix\ncm <- makeContrasts(dox_wt = wt.dox - wt.pbs,\n                    dox_top2b = top2b.dox - top2b.pbs,\n                    interaction = (top2b.dox - top2b.pbs) - (wt.dox - wt.pbs),\n                    levels = design)\n\n# View the contrasts matrix\ncm\n\n# Fit the model\nfit <- lmFit(eset, design)\n\n# Fit the contrasts\nfit2 <- contrasts.fit(fit, contrasts = cm)\n\n# Calculate the t-statistics for the contrasts\nfit2 <- eBayes(fit2)\n\n# Summarize results\nresults <- decideTests(fit2)\nsummary(results)\n\n# Create a Venn diagram\nvennDiagram(results)\n\n# Obtain the summary statistics for the contrast dox_wt\nstats_dox_wt <- topTable(fit2, coef = \"dox_wt\", number = nrow(fit2),\n                         sort.by = \"none\")\n# Obtain the summary statistics for the contrast dox_top2b\nstats_dox_top2b <- topTable(fit2, coef = \"dox_top2b\", number = nrow(fit2),\n                            sort.by = \"none\")\n# Obtain the summary statistics for the contrast interaction\nstats_interaction <- topTable(fit2, coef = \"interaction\", number = nrow(fit2),\n                              sort.by = \"none\")\n\n# Create histograms of the p-values for each contrast\nhist(stats_dox_wt[, 'P.Value'])\nhist(stats_dox_top2b[, 'P.Value'])\nhist(stats_interaction[, 'P.Value'])\n\n# Extract the gene symbols\ngene_symbols <- fit2$genes[, 'symbol']\n\n# Create a volcano plot for the contrast dox_wt\nvolcanoplot(fit2, coef = 'dox_wt', highlight = 5, names = gene_symbols)\n\n# Create a volcano plot for the contrast dox_top2b\nvolcanoplot(fit2, coef = 'dox_top2b', highlight = 5, names = gene_symbols)\n\n# Create a volcano plot for the contrast interaction\nvolcanoplot(fit2, coef = 'interaction', highlight = 5, names = gene_symbols)\n\n# Extract the entrez gene IDs\nentrez <- fit2$genes[, 'entrez']\n\n# Test for enriched KEGG Pathways for contrast dox_wt\nenrich_dox_wt <- kegga(fit2, coef = 'dox_wt', geneid = entrez, species = 'Mm')\n\n# View the top 5 enriched KEGG pathways\ntopKEGG(enrich_dox_wt, number = 5)\n\n# Test for enriched KEGG Pathways for contrast interaction\nenrich_interaction <- kegga(fit2, coef = 'interaction', geneid = entrez, species = 'Mm')\n\n# View the top 5 enriched KEGG pathways\ntopKEGG(enrich_interaction, number = 5)","metadata":{},"cell_type":"code","id":"5e4c4c11-a143-496e-a9f2-0e62fe7bb7e7","execution_count":null,"outputs":[]},{"source":"print(reads)\n\n# Get the *start* position of the first read\nstart_first <- start(reads)[1]\n\n# Get the *end* position of the last read\nend_last <- end(reads)[length(reads)]\n\n# Compute the number of reads covering each position in the selected region\ncvg <- coverage(reads)\n\n# Print a summary of the 'peaks' object\nprint(peaks)\n\n# Use the score function to find the index of the highest scoring peak\nmax_idx <- which.max(score(peaks))\n\n# Extract the genomic coordinates of the highest scoring peak using the `chrom` and `ranges` functions\nmax_peak_chrom <- chrom(peaks)[max_idx]\nmax_peak_range <- ranges(peaks)[max_idx]\n\n# Create a vector of colors to label groups (there are 2 samples per group)\ngroup <- c(primary = rep(\"blue\", 2), TURP = rep(\"red\", 2))\n\n# Plot the sample correlation matrix `sample_cor` as a heat map\n# Use the group colors to label the rows and columns of the heat map\nheatmap(sample_cor, ColSideColors = group, RowSideColors = group, \n        cexCol = 0.75, cexRow = 0.75, symm = TRUE)\n\n# Create a heat map of peak read counts\n# Use the group colors to label the columns of the heat map\nheatmap(read_counts, ColSideColors = group, labRow = \"\", cexCol = 0.75)\n\n\n### The first, ar_sets, has a list of all genes associated with peaks in the primary and treatment-resistant tumors. The second, db_sets, is a subset of the first that contains only genes associated with peaks that show evidence of differential binding between the two conditions. ###\n\n# Take a look at the full gene sets\nprint(ar_sets)\n\n# Visualise the overlap between the two groups using the `upset` function\nupset(fromList(ar_sets))\n\n# Print the genes with differential binding\nprint(db_sets)\n\n# Visualise the overlap of differentially bound peaks between the two groups using the `upset` function\nupset(fromList(db_sets))\n\n# Load reads form chr20_bam file\nreads <- readGAlignments(chr20_bam)\n\n# Create a `BamViews` object for the range 29805000 - 29820000 on chromosome 20\nbam_views <- BamViews(chr20_bam, bamRanges=GRanges(\"chr20\", IRanges(start=29805000, end=29820000)))\n\n# Load only the reads in that view\nreads_sub <- readGAlignments(bam_views)\n\n# Inspect the `reads_sub` object\nstr(reads_sub)\n\n# Load peak calls from chr20_peaks\npeaks <- import.bed(chr20_peaks)\n\n# Create a BamViews object\nbam_views <- BamViews(chr20_bam, bamRanges=peaks)\n\n# Load the reads\nreads <- readGAlignments(bam_views)\n\n# Create annotation track\npeak_track <- AnnotationTrack(peak_calls, name=\"Peaks\")\n\n# Create data track\ncover_track <- DataTrack(cover_ranges, window=10500, type=\"polygon\", name=\"Coverage\",\n                         fill.mountain=c(\"lighgrey\", \"lightgrey\"), col.mountain=\"grey\")\n\n# Produce plot\nplotTracks(list(ideogram, cover_track, peak_track, GenomeAxisTrack()), chromosome=\"chr20\", from=start_pos, to=end_pos)\n\n#to add gene track as usual make sure all libraries and packages are uploaded for this to work\ntx <- GeneRegionTrack(TxDb.Hsapiens.UCSC.hg19.knownGene, name=\"Genes\")\n\n# Find all overlaps between peaks and blacklisted regions\nblacklisted <- findOverlaps(peaks, blacklist.hg19, type=\"within\")\n\n# Create a plot to display read coverage together with peak calls and blacklisted regions in the selected region\ncover_track <- DataTrack(cover, window=10500, type=\"polygon\", name=\"Coverage\",\n                         fill.mountain=c(\"lighgrey\", \"lightgrey\"), col.mountain=\"grey\")\n\n# Calculate peak_track and region_track, plot plotTracks\npeak_track <- AnnotationTrack(peaks, name=\"Peaks\", fill=\"orange\")\nregion_track <- AnnotationTrack(region, name=\"Blacklist\")\nplotTracks(list(ideogram, cover_track, peak_track, region_track, GenomeAxisTrack()),\n           chromosome=\"chr21\", from=start(region)-1000, to=end(region)+1000)\n\n# Remove all blacklisted peaks\nclean_peaks <- peaks[-from(blacklisted)]\n\n# Load reads with mapping qualities by requesting the \"mapq\" entries\nreads <- readGAlignments(bam_file, param=ScanBamParam(what='mapq'))\n\n# Identify good quality alignments\nhigh_mapq <- mcols(reads)$mapq >= 20\n\n# Examine mapping quality distribution for high and low quality alignments\nboxplot(mcols(reads)$mapq ~ high_mapq, xlab=\"good quality alignments\", ylab=\"mapping quality\")\n\n# Remove low quality alignments\nreads_good <- subset(reads, high_mapq)\n\n# Extend reads to the average fragment length of 183 bp\nreads_ext <- resize(reads_gr, width = 183)\n\n# Compute coverage\ncover <- coverage(reads_ext)\n\n# Prepare read counts for plotting by organising them in data frames\npeak_scores <- data.frame(source=\"peaks\", fragments=peak_bins$score)\nbl_scores <- data.frame(source=\"blacklist\", fragments=bl_bins$score)\nbkg_scores <- data.frame(source=\"background\", fragments=bkg_bins$score)\nscores <- rbind(peak_scores, bl_scores, bkg_scores)\n\n# Create a boxplot of the read counts by bin type\nggplot(scores, aes(y=fragments, x=source)) + geom_boxplot()\n\n#should be normalized/distributed first\n# Compute the pairwise distances between samples using `dist`\ncover_dist <- dist(t(cover))\n\n# Use `hclust()` to create a dendrogram from the distance matrix\ncover_dendro <- hclust(cover_dist)\n\n# Plot the dendrogram\nplot(cover_dendro)\n\n# Print the `peaks` object\nprint(peaks)\n\n# Obtain the coordinates of the merged peaks\nmerged_peaks <- peaks$merged\n\n# Extract the number of merged peaks present in the data\npeak_count <- nrow(merged_peaks)\n\n# Create a heatmap using the `dba.plotHeatmap()` function\ndba.plotHeatmap(peaks, maxSites = peak_count, correlations = FALSE)\n\n###before diffbind, create a set of consensu peak calls ###\n\n# Examine the ar_binding object\nprint(ar_binding)\n\n# Identify the category corresponding to the tumor type contrast\ncontrast <- DBA_CONDITION\n\n# Establish the contrast to compare the two tumor types\ndba_peaks <- dba.contrast(ar_binding, categories =contrast, minMembers=2)\n\n# Examine the dba_peaks object to confirm that the contrast has been added\nprint(dba_peaks)\n\n# Examine the ar_binding object\nprint(ar_binding)\n\n# Identify the category corresponding to the tumor type contrast\ncontrast <- DBA_CONDITION\n\n# Establish the contrast to compare the two tumor types\ndba_peaks <- dba.contrast(ar_binding, categories =contrast, minMembers=2)\n\n# Examine the dba_peaks object to confirm that the contrast has been added\nprint(dba_peaks)\n\n###compare the charts and how they look###\n\n# Create a PCA plot using all peaks\ndba.plotPCA(ar_diff, DBA_CONDITION)\n\n# Create a PCA plot using only differentially bound peaks\ndba.plotPCA(ar_diff, DBA_CONDITION, contrast = 1)\n\n# Create a heatmap using all peaks\ndba.plotHeatmap(ar_diff, DBA_CONDITION, correlations = FALSE, maxSites = 440)\n\n# Create a heatmap using only differentially bound peaks\ndba.plotHeatmap(ar_diff, DBA_CONDITION, contrast = 1, correlations = FALSE)\n\n###if MA plot looks widely dispersed with negative bias, reflect back on normalization methods###\n\n# Create a heatmap using only differentially bound peaks\ndba.plotHeatmap(ar_diff, DBA_CONDITION, contrast = 1, correlations = FALSE)\n\n# Create a volcano plot\ndba.plotVolcano(ar_diff)\n\n# Create a box plot of the peak intensities\ncompare_groups <- dba.plotBox(ar_diff, notch=FALSE)\n\n# Inspect the returned p-values\nprint(compare_groups)\n\n# Extract peaks from ChIPQCexperiment object\npeak_calls <- peaks(ar_calls)\n\n# Only keep samples that passed QC\npeak_passed <- peak_calls[qc_pass]\n\n# Find overlaps between peak sets\npeaks_combined <- findOverlapsOfPeaks(peak_passed[[1]], peak_passed[[2]], peak_passed[[3]], peak_passed[[4]], maxgap=50)\n\n# Examine merged peak set\nprint(peaks_combined)\n\n# Obtain gene symbols\ngene_symbol <- select(org.Hs.eg.db, keys=human_genes$gene_id, columns=\"SYMBOL\", keytype=\"ENTREZID\")\n\n# Examine the structure of the returned annotations\nstr(gene_symbol)\n\n# Add gene symbols to gene coordinates\nhuman_genes$symbol <- gene_symbol$SYMBOL\n\n# Examine output\nprint(human_genes)\n\n### For each peak that is sufficiently close to the start site of a gene (as determined by the bindingRegion argument), this function will return detailed information on the gene in question in a data frame. This includes a column, insideFeature, indicating where the peak is located relative to the gene. ###\n\n# Annotate peaks with closest gene\npeak_anno <- annoPeaks(peaks_merged, human_genes, bindingType=\"startSite\", bindingRegion=c(-5000,5000))\n\n# How many peaks were found close to genes?\nlength(peak_anno)\n\n# Where are peaks located relative to genes?\ntable(peak_anno$insideFeature)\n\n# Create Venn diagram\ndba.plotVenn(ar_diff, mask=1:4)\n\n# Convert the matrix of called peaks into a data frame\ncalled_peaks <- as.data.frame(ar_diff$called)\n\n# Create UpSet plot\nupset(called_peaks, keep.order = TRUE, sets=colnames(ar_diff$called), order.by=\"freq\")\n\n# Plot distribution of distances between peaks and transcription start sites\nplot_dist_to_tss(peaks, genome = \"hg19\")\n\n# Plot relationship between gene length and presence of peaks\nplot_chipenrich_spline(peaks, genome = 'hg19', mappability=50)\n\n# Select all peaks with higher intensity in treatment resistant samples\nturp_peaks <- peaks_binding[, \"GSM1598218\"] + peaks_binding[, \"GSM1598219\"] < peaks_binding[, \"GSM1598223\"] + peaks_binding[, \"GSM1598225\"]\n\n# Run enrichment analysis\nenrich_turp <- chipenrich(peaks_comb[turp_peaks, ], genome=\"hg19\", \n                   genesets = \"hallmark\", out_name = NULL, \n                   locusdef = \"nearest_tss\", qc_plots=FALSE)\n\n# Print the results of the analysis\nprint(enrich_turp$results)\n\n# Examine the top gene sets\nhead(enrich_primary$results)\n\n# Extract the gene IDs for the top ranking set\ngenes <- enrich_primary$results$Geneset.Peak.Genes[1]\n\n# Split gene IDs into a vector\ngene_ids <- strsplit(genes, ', ')[[1]]\n\n# Convert gene IDs to gene symbols\ngene_symbol <- select(org.Hs.eg.db, keys=gene_ids, columns=\"SYMBOL\", keytype=\"ENTREZID\")\n\n# Print the result\nprint(gene_symbol)\n\n# This is the base URL for all KEGG pathways\nbase_url <- \"https://www.kegg.jp/pathway/\"\n\n# Add pathway ID to URL\npath_url <- paste0(base_url, top_path)\n\n# Collapse gene IDs into selection string\ngene_select <- paste(genes, collapse=\"+\")\n\n# Add gene IDs to URL\npath_url <- paste(path_url, gene_select, sep=\"+\")","metadata":{},"cell_type":"code","id":"d3fda15e-bdba-4f26-bf7d-b5f2a899c33c","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"display_name":"R","language":"R","name":"ir"},"language_info":{"name":"R","codemirror_mode":"r","pygments_lexer":"r","mimetype":"text/x-r-source","file_extension":".r","version":"4.2.1"},"editor":"DataCamp Workspace"},"nbformat":4,"nbformat_minor":5}